### ISSUE TITLE
`Invalid size request on a scalable vector.` when building Android libart.so for armv9

### ISSUE THREAD

**Issue by pirama-arumuga-nainar** [llvm/llvm-project#59480]
In https://android-review.googlesource.com/c/platform/build/soong/+/2292699, there's a build failure when compiling for armv9 with LTO.
```console
LLVM ERROR: Invalid size request on a scalable vector.
```

Please see repro at https://android-review.googlesource.com/c/platform/prebuilts/clang/host/windows-x86/+/2345751 [1].

Clang treats errors related to SVE size as warnings by passing [-mllvm,-treat-scalable-fixed-error-as-warning](https://github.com/llvm/llvm-project/blob/f45195784168cced1e12fe888d6a906b8a4b424f/clang/lib/Driver/ToolChains/Clang.cpp#L5477).  It's not clear if this flag is still needed or just a forgotten cleanup.  The `-treat-scalable-fixed-error-as-warning` workaround is not in place during LTO (as mentioned [here](https://reviews.llvm.org/D98856#2647927)) and needs to be explicitly passed to the backend via -Wl.

[1] This is in AOSP gerrit to workaround file-size issues in github/phabricator.  To fetch it, please run:
```shell
git fetch https://android.googlesource.com/platform/prebuilts/clang/host/windows-x86 refs/changes/51/2345751/1 && git checkout FETCH_HEAD
```

(@kbeyls @enh-google @stephenhines )

**Comment by pirama-arumuga-nainar** [comment]
This repros on 8482e95f75d02227fbf51527680c0b5424bacb69 from Dec 6.  The following is the relevant part of the stack trace:

```
#11 0x0000564b95589672 llvm::TypeSize::operator unsigned long() const /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/lib/Support/TypeSize.cpp:51:5
#12 0x0000564b99984a8a llvm::InstCombinerImpl::simplifyIntToPtrRoundTripCast(llvm::Value*) /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
375:23
#13 0x0000564b99a70ff9 llvm::InstCombinerImpl::foldICmpWithCastOp(llvm::ICmpInst&) /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5049:10
#14 0x0000564b99a75344 llvm::InstCombinerImpl::visitICmpInst(llvm::ICmpInst&) /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:6378:20
#15 0x0000564b999a735d llvm::InstVisitor<llvm::InstCombinerImpl, llvm::Instruction*>::visitICmp(llvm::ICmpInst&) /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/include/llvm/IR/Instruction.de
f:206:1
#16 0x0000564b9999d1c4 llvm::InstVisitor<llvm::InstCombinerImpl, llvm::Instruction*>::visit(llvm::Instruction&) /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/include/llvm/IR/Instruction.def
:206:1
#17 0x0000564b99995ea3 llvm::InstCombinerImpl::run() /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:4309:22
#18 0x0000564b99997634 combineInstructionsOverFunction(llvm::Function&, llvm::InstructionWorklist&, llvm::AAResults*, llvm::AssumptionCache&, llvm::TargetLibraryInfo&, llvm::TargetTransformInfo&, llvm::Dominato
rTree&, llvm::OptimizationRemarkEmitter&, llvm::BlockFrequencyInfo*, llvm::ProfileSummaryInfo*, unsigned int, llvm::LoopInfo*) /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/lib/Transforms/I
nstCombine/InstructionCombining.cpp:4604:9
#19 0x0000564b9999706c llvm::InstCombinePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) /usr/local/google/home/pirama/work/llvm-upstream/llvm-project/llvm/lib/Transforms/InstCombine/Instructi
onCombining.cpp:4635:7
```

**Comment by pirama-arumuga-nainar** [comment]
The crash is fixed with [D139911](https://reviews.llvm.org/D139911).  Thanks Craig!

**Comment by topperc** [comment]
New patch  https://reviews.llvm.org/D139962

**Comment by topperc** [comment]
Fixed by d08e3ae271a9748af6c99be7e455730fb110b0a5

**Comment by pirama-arumuga-nainar** [comment]
Thanks for the quick fix @topperc!

Any thoughts on whether the `-treat-scalable-fixed-error-as-warning` flag is still needed in the backend?

**Comment by pirama-arumuga-nainar** [comment]
> Any thoughts on whether the `-treat-scalable-fixed-error-as-warning` flag is still needed in the backend?

Also adding @paulwalker-arm for his thoughts on the above query.

**Comment by paulwalker-arm** [comment]
We've likely hit a critical mass of scalable vector enablement whereby we perhaps don't need to downgrade such errors to warnings by default anymore.  In fact, this might help drag the enablement to completion if it means getting more bug reports.  Those willing to take a chance can still add the option to their build flags to unblock themselves. That said, I'd be a little nervous about making such a change this close to the next release so perhaps something for after the next branch is cut?

**Comment by pirama-arumuga-nainar** [comment]
clang-17 is now branched.  Can we turn the warnings to errors now?

**Comment by paulwalker-arm** [comment]
I've created https://reviews.llvm.org/D146146 to restore the warning to an error.
